osu.Game.Rulesets.Taiko/Mods/TaikoModAlternate.cs

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Linq;
+using osu.Framework.Bindables;
+using osu.Game.Configuration;
+using osu.Game.Rulesets.Mods;
+using osu.Game.Rulesets.Objects.Types;
+using osu.Game.Rulesets.Scoring;
+using osu.Game.Rulesets.Taiko.Objects;
+
+namespace osu.Game.Rulesets.Taiko.Mods
+{
+    public class TaikoModAlternate : ModAlternate<TaikoHitObject, TaikoAction>
+    {
+        [SettingSource("Playstyle", "Preferred Alternate Playstyle")]
+        public Bindable<Playstyle> Style { get; } = new Bindable<Playstyle>();
+
+        private TaikoAction? lastAction1;
+        private TaikoAction? lastAction2;
+        private TaikoAction lastAction;
+        private TaikoHitObject lastHitObject;
+        private double lastActionTime = 0;
+        private const double strong_hit_window = 30;
+
+        protected override void ResetActionStates()
+        {
+            lastAction1 = lastAction2 = null;
+        }
+
+        protected override bool OnPressed(TaikoAction action)
+        {
+            var blockInput = false;
+
+            var hitObject = HitObjects.FirstOrDefault((h) =>
+            {
+                var time = Interceptor.Time.Current;
+                var window = h.HitWindows.WindowFor(HitResult.Miss);
+                var endTime = (h as IHasDuration)?.EndTime ?? h.StartTime;
+                return (time > h.StartTime - window) && (time < endTime + window);
+            });
+
+            if (hitObject?.IsStrong ?? false)
+            {
+                if ((hitObject == lastHitObject) && (Interceptor.Time.Current - lastActionTime <= strong_hit_window))
+                {
+                    if ((lastAction == TaikoAction.LeftRim) && (action == TaikoAction.RightRim) ||
+                        (lastAction == TaikoAction.RightRim) && (action == TaikoAction.LeftRim) ||
+                        (lastAction == TaikoAction.LeftCentre) && (action == TaikoAction.RightCentre) ||
+                        (lastAction == TaikoAction.RightCentre) && (action == TaikoAction.LeftCentre))
```

Comment:
These parens are very weird to me; they don't really help me read this (I *sort of* remember operator precedence rules for `&&` and `||` but it would be more readable if they were explicit). Would probably change to
```suggestion
                    if ((lastAction == TaikoAction.LeftRim && action == TaikoAction.RightRim) ||
                        (lastAction == TaikoAction.RightRim && action == TaikoAction.LeftRim) ||
                        (lastAction == TaikoAction.LeftCentre && action == TaikoAction.RightCentre) ||
                        (lastAction == TaikoAction.RightCentre && action == TaikoAction.LeftCentre))
```

(You could also use a `compareUnordered` here which I wrote a version of in the other suggestion.)